File: backend/src/controllers/admin.controller.js
const User = require('../models/User');
const HealthRecord = require('../models/HealthRecord');
const SubscriptionPlan = require('../models/SubscriptionPlan');
const UserSubscription = require('../models/UserSubscription');

function monthsFromNow(n) {
  const d = new Date();
  d.setMonth(d.getMonth() + n);
  return d;
}

function yearsFromNow(n) {
  const d = new Date();
  d.setFullYear(d.getFullYear() + n);
  return d;
}

function endDateForPlan(plan, start = new Date()) {
  if (!plan) return null;
  if (plan.interval === 'lifetime') return null;
  if (plan.interval === 'yearly') return yearsFromNow(1);
  return monthsFromNow(1);
}

/** ---------- Overview ---------- */
exports.getOverview = async (_req, res) => {
  try {
    const [users, plans, subscriptions] = await Promise.all([
      User.find().select('role isActive hasDiabetes gender'),
      SubscriptionPlan.find({ isActive: true }),
      UserSubscription.find().populate('plan', 'name code price currency interval'),
    ]);

    const counts = {
      users: users.length,
      activeUsers: users.filter((u) => u.isActive).length,
      admin: users.filter((u) => u.role === 'admin').length,
      doctor: users.filter((u) => u.role === 'doctor').length,
      patient: users.filter((u) => u.role === 'patient').length,
      diabetes: users.filter((u) => u.hasDiabetes).length,
      women: users.filter((u) => u.gender === 'woman').length,
      men: users.filter((u) => u.gender === 'man').length,
      plans: plans.length,
      activeSubscriptions: subscriptions.filter((s) => s.status === 'active' || s.status === 'trial')
        .length,
    };

    const mrr = subscriptions
      .filter((s) => (s.status === 'active' || s.status === 'trial') && s.plan)
      .reduce((sum, s) => {
        const price = s.plan.price || 0;
        if (s.plan.interval === 'yearly') return sum + price / 12;
        if (s.plan.interval === 'lifetime') return sum;
        return sum + price;
      }, 0);

    res.json({ counts, mrr: Math.round(mrr * 100) / 100, currency: 'MAD' });
  } catch (error) {
    console.error('admin overview:', error);
    res.status(500).json({ message: 'Failed to load overview' });
  }
};

/** ---------- Users ---------- */
exports.listUsers = async (req, res) => {
  try {
    const includeInactive = req.query.all === '1';
    const filter = includeInactive ? {} : { isActive: true };
    const users = await User.find(filter)
      .select(
        'firstName lastName email role gender hasDiabetes diabetesType specialty phone assignedDoctor isActive createdAt'
      )
      .populate('assignedDoctor', 'firstName lastName email')
      .sort({ createdAt: -1 });

    const subs = await UserSubscription.find({
      user: { $in: users.map((u) => u._id) },
      status: { $in: ['active', 'trial'] },
    }).populate('plan', 'name code price currency interval');

    const subByUser = Object.fromEntries(subs.map((s) => [String(s.user), s]));

    res.json({
      users: users.map((u) => ({
        ...u.toObject(),
        subscription: subByUser[String(u._id)] || null,
      })),
    });
  } catch (error) {
    console.error('admin listUsers:', error);
    res.status(500).json({ message: 'Failed to load users' });
  }
};

exports.createUser = async (req, res) => {
  try {
    const {
      firstName,
      lastName,
      email,
      password,
      role,
      gender,
      phone,
      specialty,
      hasDiabetes,
      diabetesType,
      assignedDoctor,
      isActive,
    } = req.body;

    if (!firstName || !lastName || !email || !password) {
      return res.status(400).json({ message: 'firstName, lastName, email, password required' });
    }

    const exists = await User.findOne({ email: email.toLowerCase() });
    if (exists) return res.status(409).json({ message: 'Email already registered' });

    const safeRole = ['patient', 'doctor', 'admin'].includes(role) ? role : 'patient';
    const user = await User.create({
      firstName,
      lastName,
      email,
      password,
      role: safeRole,
      gender: gender === 'man' ? 'man' : 'woman',
      phone: phone || '',
      specialty: safeRole === 'doctor' ? specialty || 'General medicine' : '',
      hasDiabetes: Boolean(hasDiabetes),
      diabetesType: hasDiabetes ? diabetesType || 'type2' : 'none',
      assignedDoctor: assignedDoctor || null,
      isActive: isActive !== false,
    });

    if (user.role === 'patient') {
      await HealthRecord.create({ patient: user._id });
    }

    res.status(201).json({ user: user.toSafeJSON() });
  } catch (error) {
    console.error('admin createUser:', error);
    res.status(500).json({ message: 'Failed to create user' });
  }
};

exports.updateUser = async (req, res) => {
  try {
    const user = await User.findById(req.params.id).select('+password');
    if (!user) return res.status(404).json({ message: 'User not found' });

    const {
      firstName,
      lastName,
      email,
      role,
      gender,
      phone,
      specialty,
      hasDiabetes,
      diabetesType,
      assignedDoctor,
      isActive,
      password,
    } = req.body;

    if (firstName !== undefined) user.firstName = firstName;
    if (lastName !== undefined) user.lastName = lastName;
    if (email !== undefined) user.email = email.toLowerCase();
    if (role !== undefined && ['patient', 'doctor', 'admin'].includes(role)) user.role = role;
    if (gender !== undefined) user.gender = gender === 'man' ? 'man' : 'woman';
    if (phone !== undefined) user.phone = phone;
    if (specialty !== undefined) user.specialty = specialty;
    if (hasDiabetes !== undefined) user.hasDiabetes = Boolean(hasDiabetes);
    if (diabetesType !== undefined) user.diabetesType = diabetesType;
    if (assignedDoctor !== undefined) user.assignedDoctor = assignedDoctor || null;
    if (isActive !== undefined) user.isActive = Boolean(isActive);
    if (password && String(password).length >= 6) user.password = password;

    // Prevent locking yourself out
    if (String(user._id) === String(req.user._id) && user.isActive === false) {
      return res.status(400).json({ message: 'You cannot deactivate your own account' });
    }
    if (String(user._id) === String(req.user._id) && user.role !== 'admin') {
      return res.status(400).json({ message: 'You cannot remove your own admin role' });
    }

    await user.save();
    res.json({ user: user.toSafeJSON() });
  } catch (error) {
    console.error('admin updateUser:', error);
    if (error.code === 11000) {
      return res.status(409).json({ message: 'Email already in use' });
    }
    res.status(500).json({ message: 'Failed to update user' });
  }
};

exports.deleteUser = async (req, res) => {
  try {
    if (String(req.params.id) === String(req.user._id)) {
      return res.status(400).json({ message: 'You cannot delete your own account' });
    }

    const user = await User.findById(req.params.id);
    if (!user) return res.status(404).json({ message: 'User not found' });

    user.isActive = false;
    await user.save();

    await UserSubscription.updateMany(
      { user: user._id, status: { $in: ['active', 'trial'] } },
      { status: 'cancelled', autoRenew: false }
    );

    res.json({ message: 'User deactivated', user: user.toSafeJSON() });
  } catch (error) {
    console.error('admin deleteUser:', error);
    res.status(500).json({ message: 'Failed to deactivate user' });
  }
};

/** ---------- Plans ---------- */
exports.listPlans = async (_req, res) => {
  try {
    const plans = await SubscriptionPlan.find().sort({ sortOrder: 1, price: 1 });
    res.json({ plans });
  } catch (error) {
    res.status(500).json({ message: 'Failed to load plans' });
  }
};

exports.createPlan = async (req, res) => {
  try {
    const plan = await SubscriptionPlan.create(req.body);
    res.status(201).json({ plan });
  } catch (error) {
    console.error('admin createPlan:', error);
    if (error.code === 11000) {
      return res.status(409).json({ message: 'Plan code already exists' });
    }
    res.status(500).json({ message: 'Failed to create plan' });
  }
};

exports.updatePlan = async (req, res) => {
  try {
    const plan = await SubscriptionPlan.findByIdAndUpdate(req.params.id, req.body, {
      new: true,
      runValidators: true,
    });
    if (!plan) return res.status(404).json({ message: 'Plan not found' });
    res.json({ plan });
  } catch (error) {
    console.error('admin updatePlan:', error);
    res.status(500).json({ message: 'Failed to update plan' });
  }
};

exports.deletePlan = async (req, res) => {
  try {
    const plan = await SubscriptionPlan.findById(req.params.id);
    if (!plan) return res.status(404).json({ message: 'Plan not found' });

    const activeSubs = await UserSubscription.countDocuments({
      plan: plan._id,
      status: { $in: ['active', 'trial'] },
    });
    if (activeSubs > 0) {
      plan.isActive = false;
      await plan.save();
      return res.json({
        plan,
        message: 'Plan has active subscribers — marked inactive instead of deleted',
      });
    }

    await plan.deleteOne();
    res.json({ message: 'Plan deleted' });
  } catch (error) {
    res.status(500).json({ message: 'Failed to delete plan' });
  }
};

/** ---------- Subscriptions ---------- */
exports.listSubscriptions = async (_req, res) => {
  try {
    const subscriptions = await UserSubscription.find()
      .populate('user', 'firstName lastName email role isActive')
      .populate('plan', 'name code price currency interval')
      .populate('managedBy', 'firstName lastName')
      .sort({ updatedAt: -1 });
    res.json({ subscriptions });
  } catch (error) {
    res.status(500).json({ message: 'Failed to load subscriptions' });
  }
};

exports.assignSubscription = async (req, res) => {
  try {
    const { userId, planId, status, endDate, autoRenew, notes, startDate } = req.body;
    if (!userId || !planId) {
      return res.status(400).json({ message: 'userId and planId are required' });
    }

    const [user, plan] = await Promise.all([
      User.findById(userId),
      SubscriptionPlan.findById(planId),
    ]);
    if (!user) return res.status(404).json({ message: 'User not found' });
    if (!plan) return res.status(404).json({ message: 'Plan not found' });

    // Cancel previous active/trial
    await UserSubscription.updateMany(
      { user: userId, status: { $in: ['active', 'trial'] } },
      { status: 'cancelled', autoRenew: false }
    );

    const start = startDate ? new Date(startDate) : new Date();
    const sub = await UserSubscription.create({
      user: userId,
      plan: planId,
      status: ['active', 'trial', 'cancelled', 'expired', 'past_due'].includes(status)
        ? status
        : 'active',
      startDate: start,
      endDate: endDate ? new Date(endDate) : endDateForPlan(plan, start),
      autoRenew: autoRenew !== false,
      notes: notes || '',
      managedBy: req.user._id,
    });

    const populated = await UserSubscription.findById(sub._id)
      .populate('user', 'firstName lastName email role')
      .populate('plan', 'name code price currency interval');

    res.status(201).json({ subscription: populated });
  } catch (error) {
    console.error('admin assignSubscription:', error);
    res.status(500).json({ message: 'Failed to assign subscription' });
  }
};

exports.updateSubscription = async (req, res) => {
  try {
    const sub = await UserSubscription.findById(req.params.id);
    if (!sub) return res.status(404).json({ message: 'Subscription not found' });

    const { planId, status, endDate, autoRenew, notes, startDate } = req.body;

    if (planId) {
      const plan = await SubscriptionPlan.findById(planId);
      if (!plan) return res.status(404).json({ message: 'Plan not found' });
      sub.plan = planId;
      if (!endDate && !sub.endDate) sub.endDate = endDateForPlan(plan);
    }
    if (status !== undefined) sub.status = status;
    if (endDate !== undefined) sub.endDate = endDate ? new Date(endDate) : null;
    if (startDate !== undefined) sub.startDate = new Date(startDate);
    if (autoRenew !== undefined) sub.autoRenew = Boolean(autoRenew);
    if (notes !== undefined) sub.notes = notes;
    sub.managedBy = req.user._id;

    await sub.save();

    const populated = await UserSubscription.findById(sub._id)
      .populate('user', 'firstName lastName email role')
      .populate('plan', 'name code price currency interval');

    res.json({ subscription: populated });
  } catch (error) {
    console.error('admin updateSubscription:', error);
    res.status(500).json({ message: 'Failed to update subscription' });
  }
};

exports.cancelSubscription = async (req, res) => {
  try {
    const sub = await UserSubscription.findById(req.params.id);
    if (!sub) return res.status(404).json({ message: 'Subscription not found' });

    sub.status = 'cancelled';
    sub.autoRenew = false;
    sub.managedBy = req.user._id;
    if (req.body.notes) sub.notes = req.body.notes;
    await sub.save();

    const populated = await UserSubscription.findById(sub._id)
      .populate('user', 'firstName lastName email role')
      .populate('plan', 'name code price currency interval');

    res.json({ subscription: populated });
  } catch (error) {
    res.status(500).json({ message: 'Failed to cancel subscription' });
  }
};

File: backend-dotnet/MyHeath.Api/Controllers/AdminController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using MongoDB.Driver;
using MyHeath.Api.Infrastructure;
using MyHeath.Api.Models;
using MyHeath.Api.Services;

namespace MyHeath.Api.Controllers;

[ApiController]
[Route("api/admin")]
[Authorize(Roles = Roles.Admin)]
public class AdminController : ApiControllerBase
{
    private readonly AuditService _audit;

    public AdminController(MongoContext db, AuditService audit) : base(db) => _audit = audit;

    [HttpGet("overview")]
    public async Task<IActionResult> Overview()
    {
        var users = await Db.Users.Find(_ => true).ToListAsync();
        var plans = await Db.Plans.Find(p => p.IsActive).ToListAsync();
        var subs = await Db.Subscriptions.Find(_ => true).ToListAsync();
        var planMap = (await Db.Plans.Find(_ => true).ToListAsync()).ToDictionary(p => p.Id);

        var activeSubs = subs.Where(s => s.Status is "active" or "trial").ToList();
        double mrr = 0;
        foreach (var s in activeSubs)
        {
            if (!planMap.TryGetValue(s.Plan, out var plan)) continue;
            mrr += plan.Interval == "yearly" ? plan.Price / 12 : plan.Interval == "lifetime" ? 0 : plan.Price;
        }

        return Ok(new
        {
            counts = new
            {
                users = users.Count,
                activeUsers = users.Count(u => u.IsActive),
                admin = users.Count(u => u.Role == Roles.Admin),
                doctor = users.Count(u => u.Role == Roles.Doctor),
                patient = users.Count(u => u.Role == Roles.Patient),
                diabetes = users.Count(u => u.HasDiabetes),
                women = users.Count(u => u.Gender == "woman"),
                men = users.Count(u => u.Gender == "man"),
                plans = plans.Count,
                activeSubscriptions = activeSubs.Count,
                appointments = await Db.Appointments.CountDocumentsAsync(_ => true),
                notifications = await Db.Notifications.CountDocumentsAsync(_ => true)
            },
            mrr = Math.Round(mrr, 2),
            currency = "MAD"
        });
    }

    [HttpGet("users")]
    public async Task<IActionResult> ListUsers([FromQuery] string? all)
    {
        var filter = all == "1" ? Builders<User>.Filter.Empty : Builders<User>.Filter.Eq(u => u.IsActive, true);
        var users = await Db.Users.Find(filter).SortByDescending(u => u.CreatedAt).ToListAsync();
        var subs = await Db.Subscriptions.Find(s => s.Status == "active" || s.Status == "trial").ToListAsync();
        var plans = (await Db.Plans.Find(_ => true).ToListAsync()).ToDictionary(p => p.Id);
        var subByUser = subs.GroupBy(s => s.User).ToDictionary(g => g.Key, g => g.First());

        return Ok(new
        {
            users = users.Select(u =>
            {
                object? subscription = null;
                if (subByUser.TryGetValue(u.Id, out var sub) && plans.TryGetValue(sub.Plan, out var plan))
                {
                    subscription = new
                    {
                        _id = sub.Id,
                        status = sub.Status,
                        plan = new { _id = plan.Id, name = plan.Name, code = plan.Code, price = plan.Price, currency = plan.Currency, interval = plan.Interval }
                    };
                }
                return new
                {
                    _id = u.Id,
                    firstName = u.FirstName,
                    lastName = u.LastName,
                    email = u.Email,
                    role = u.Role,
                    gender = u.Gender,
                    hasDiabetes = u.HasDiabetes,
                    diabetesType = u.DiabetesType,
                    specialty = u.Specialty,
                    phone = u.Phone,
                    assignedDoctor = u.AssignedDoctor,
                    isActive = u.IsActive,
                    createdAt = u.CreatedAt,
                    subscription
                };
            })
        });
    }

    public record UserUpsertDto(
        string? FirstName, string? LastName, string? Email, string? Password, string? Role,
        string? Gender, string? Phone, string? Specialty, bool? HasDiabetes, string? DiabetesType,
        string? AssignedDoctor, bool? IsActive);

    [HttpPost("users")]
    public async Task<IActionResult> CreateUser([FromBody] UserUpsertDto dto)
    {
        var admin = await CurrentUserAsync();
        if (string.IsNullOrWhiteSpace(dto.FirstName) || string.IsNullOrWhiteSpace(dto.Email) || string.IsNullOrWhiteSpace(dto.Password))
            return Fail("firstName, email, password required");
        var email = dto.Email.Trim().ToLowerInvariant();
        if (await Db.Users.Find(u => u.Email == email).AnyAsync()) return Fail("Email already registered", 409);

        var role = dto.Role is "patient" or "doctor" or "admin" ? dto.Role! : Roles.Patient;
        var user = new User
        {
            FirstName = dto.FirstName!,
            LastName = dto.LastName ?? "",
            Email = email,
            Password = BCrypt.Net.BCrypt.HashPassword(dto.Password!, 12),
            Role = role,
            Gender = dto.Gender == "man" ? "man" : "woman",
            Phone = dto.Phone ?? "",
            Specialty = role == Roles.Doctor ? (dto.Specialty ?? "General medicine") : "",
            HasDiabetes = dto.HasDiabetes == true,
            DiabetesType = dto.HasDiabetes == true ? (dto.DiabetesType ?? "type2") : "none",
            AssignedDoctor = string.IsNullOrEmpty(dto.AssignedDoctor) ? null : dto.AssignedDoctor,
            IsActive = dto.IsActive != false
        };
        await Db.Users.InsertOneAsync(user);
        if (user.Role == Roles.Patient)
            await Db.HealthRecords.InsertOneAsync(new HealthRecord { Patient = user.Id });

        await _audit.LogAsync(admin!.Id, admin.Email, "create", "user", user.Id, email);
        return StatusCode(201, new { user = UserMapper.ToSafe(user) });
    }

    [HttpPut("users/{id}")]
    public async Task<IActionResult> UpdateUser(string id, [FromBody] UserUpsertDto dto)
    {
        var admin = await CurrentUserAsync();
        var user = await Db.Users.Find(u => u.Id == id).FirstOrDefaultAsync();
        if (user is null) return Fail("User not found", 404);
        if (dto.FirstName != null) user.FirstName = dto.FirstName;
        if (dto.LastName != null) user.LastName = dto.LastName;
        if (dto.Email != null) user.Email = dto.Email.ToLowerInvariant();
        if (dto.Role is "patient" or "doctor" or "admin") user.Role = dto.Role;
        if (dto.Gender != null) user.Gender = dto.Gender == "man" ? "man" : "woman";
        if (dto.Phone != null) user.Phone = dto.Phone;
        if (dto.Specialty != null) user.Specialty = dto.Specialty;
        if (dto.HasDiabetes.HasValue) user.HasDiabetes = dto.HasDiabetes.Value;
        if (dto.DiabetesType != null) user.DiabetesType = dto.DiabetesType;
        if (dto.AssignedDoctor != null) user.AssignedDoctor = string.IsNullOrEmpty(dto.AssignedDoctor) ? null : dto.AssignedDoctor;
        if (dto.IsActive.HasValue) user.IsActive = dto.IsActive.Value;
        if (!string.IsNullOrEmpty(dto.Password) && dto.Password.Length >= 6)
            user.Password = BCrypt.Net.BCrypt.HashPassword(dto.Password, 12);
        if (user.Id == admin!.Id && (!user.IsActive || user.Role != Roles.Admin))
            return Fail("You cannot remove your own admin access");

        user.UpdatedAt = DateTime.UtcNow;
        await Db.Users.ReplaceOneAsync(u => u.Id == id, user);
        await _audit.LogAsync(admin.Id, admin.Email, "update", "user", id, user.Email);
        return Ok(new { user = UserMapper.ToSafe(user) });
    }

    [HttpDelete("users/{id}")]
    public async Task<IActionResult> DeleteUser(string id)
    {
        var admin = await CurrentUserAsync();
        if (id == admin!.Id) return Fail("You cannot delete your own account");
        var user = await Db.Users.Find(u => u.Id == id).FirstOrDefaultAsync();
        if (user is null) return Fail("User not found", 404);
        user.IsActive = false;
        user.UpdatedAt = DateTime.UtcNow;
        await Db.Users.ReplaceOneAsync(u => u.Id == id, user);
        await Db.Subscriptions.UpdateManyAsync(
            s => s.User == id && (s.Status == "active" || s.Status == "trial"),
            Builders<UserSubscription>.Update.Set(s => s.Status, "cancelled").Set(s => s.AutoRenew, false));
        await _audit.LogAsync(admin.Id, admin.Email, "deactivate", "user", id, user.Email);
        return Ok(new { message = "User deactivated", user = UserMapper.ToSafe(user) });
    }

    [HttpGet("plans")]
    public async Task<IActionResult> ListPlans() =>
        Ok(new { plans = await Db.Plans.Find(_ => true).SortBy(p => p.SortOrder).ToListAsync() });

    [HttpPost("plans")]
    public async Task<IActionResult> CreatePlan([FromBody] SubscriptionPlan plan)
    {
        plan.Id = MongoDB.Bson.ObjectId.GenerateNewId().ToString();
        plan.Code = (plan.Code ?? "").ToUpperInvariant();
        plan.CreatedAt = DateTime.UtcNow;
        await Db.Plans.InsertOneAsync(plan);
        return StatusCode(201, new { plan });
    }

    [HttpPut("plans/{id}")]
    public async Task<IActionResult> UpdatePlan(string id, [FromBody] SubscriptionPlan body)
    {
        var plan = await Db.Plans.Find(p => p.Id == id).FirstOrDefaultAsync();
        if (plan is null) return Fail("Plan not found", 404);
        body.Id = id;
        body.Code = plan.Code;
        await Db.Plans.ReplaceOneAsync(p => p.Id == id, body);
        return Ok(new { plan = body });
    }

    [HttpDelete("plans/{id}")]
    public async Task<IActionResult> DeletePlan(string id)
    {
        var plan = await Db.Plans.Find(p => p.Id == id).FirstOrDefaultAsync();
        if (plan is null) return Fail("Plan not found", 404);
        var active = await Db.Subscriptions.CountDocumentsAsync(s => s.Plan == id && (s.Status == "active" || s.Status == "trial"));
        if (active > 0)
        {
            plan.IsActive = false;
            await Db.Plans.ReplaceOneAsync(p => p.Id == id, plan);
            return Ok(new { plan, message = "Plan has active subscribers — marked inactive instead of deleted" });
        }
        await Db.Plans.DeleteOneAsync(p => p.Id == id);
        return Ok(new { message = "Plan deleted" });
    }

    [HttpGet("subscriptions")]
    public async Task<IActionResult> ListSubscriptions()
    {
        var subs = await Db.Subscriptions.Find(_ => true).SortByDescending(s => s.UpdatedAt).ToListAsync();
        var users = (await Db.Users.Find(_ => true).ToListAsync()).ToDictionary(u => u.Id);
        var plans = (await Db.Plans.Find(_ => true).ToListAsync()).ToDictionary(p => p.Id);
        return Ok(new
        {
            subscriptions = subs.Select(s => new
            {
                _id = s.Id,
                status = s.Status,
                startDate = s.StartDate,
                endDate = s.EndDate,
                autoRenew = s.AutoRenew,
                notes = s.Notes,
                user = users.TryGetValue(s.User, out var u)
                    ? new { _id = u.Id, firstName = u.FirstName, lastName = u.LastName, email = u.Email, role = u.Role, isActive = u.IsActive }
                    : null,
                plan = plans.TryGetValue(s.Plan, out var p)
                    ? new { _id = p.Id, name = p.Name, code = p.Code, price = p.Price, currency = p.Currency, interval = p.Interval }
                    : null
            })
        });
    }

    public record AssignSubDto(string UserId, string PlanId, string? Status, DateTime? EndDate, bool? AutoRenew, string? Notes, DateTime? StartDate);

    [HttpPost("subscriptions")]
    public async Task<IActionResult> AssignSubscription([FromBody] AssignSubDto dto)
    {
        var admin = await CurrentUserAsync();
        var user = await Db.Users.Find(u => u.Id == dto.UserId).FirstOrDefaultAsync();
        var plan = await Db.Plans.Find(p => p.Id == dto.PlanId).FirstOrDefaultAsync();
        if (user is null || plan is null) return Fail("User or plan not found", 404);

        await Db.Subscriptions.UpdateManyAsync(
            s => s.User == dto.UserId && (s.Status == "active" || s.Status == "trial"),
            Builders<UserSubscription>.Update.Set(s => s.Status, "cancelled").Set(s => s.AutoRenew, false));

        var start = dto.StartDate ?? DateTime.UtcNow;
        DateTime? end = dto.EndDate;
        if (end is null && plan.Interval != "lifetime")
            end = plan.Interval == "yearly" ? start.AddYears(1) : start.AddMonths(1);

        var sub = new UserSubscription
        {
            User = dto.UserId,
            Plan = dto.PlanId,
            Status = dto.Status is "active" or "trial" or "cancelled" or "expired" or "past_due" ? dto.Status! : "active",
            StartDate = start,
            EndDate = end,
            AutoRenew = dto.AutoRenew != false,
            Notes = dto.Notes ?? "",
            ManagedBy = admin!.Id
        };
        await Db.Subscriptions.InsertOneAsync(sub);
        await _audit.LogAsync(admin.Id, admin.Email, "assign_subscription", "subscription", sub.Id, $"{user.Email} -> {plan.Code}");
        return StatusCode(201, new { subscription = sub });
    }

    [HttpPut("subscriptions/{id}")]
    public async Task<IActionResult> UpdateSubscription(string id, [FromBody] AssignSubDto dto)
    {
        var sub = await Db.Subscriptions.Find(s => s.Id == id).FirstOrDefaultAsync();
        if (sub is null) return Fail("Subscription not found", 404);
        if (!string.IsNullOrEmpty(dto.PlanId)) sub.Plan = dto.PlanId;
        if (!string.IsNullOrEmpty(dto.Status)) sub.Status = dto.Status!;
        if (dto.EndDate.HasValue) sub.EndDate = dto.EndDate;
        if (dto.StartDate.HasValue) sub.StartDate = dto.StartDate.Value;
        if (dto.AutoRenew.HasValue) sub.AutoRenew = dto.AutoRenew.Value;
        if (dto.Notes != null) sub.Notes = dto.Notes;
        sub.UpdatedAt = DateTime.UtcNow;
        await Db.Subscriptions.ReplaceOneAsync(s => s.Id == id, sub);
        return Ok(new { subscription = sub });
    }

    [HttpPost("subscriptions/{id}/cancel")]
    public async Task<IActionResult> CancelSubscription(string id, [FromBody] AssignSubDto? dto)
    {
        var admin = await CurrentUserAsync();
        var sub = await Db.Subscriptions.Find(s => s.Id == id).FirstOrDefaultAsync();
        if (sub is null) return Fail("Subscription not found", 404);
        sub.Status = "cancelled";
        sub.AutoRenew = false;
        if (dto?.Notes != null) sub.Notes = dto.Notes;
        sub.ManagedBy = admin!.Id;
        sub.UpdatedAt = DateTime.UtcNow;
        await Db.Subscriptions.ReplaceOneAsync(s => s.Id == id, sub);
        return Ok(new { subscription = sub });
    }

    [HttpGet("audit")]
    public async Task<IActionResult> Audit([FromQuery] int limit = 50)
    {
        var logs = await Db.AuditLogs.Find(_ => true).SortByDescending(a => a.CreatedAt)
            .Limit(Math.Clamp(limit, 1, 200)).ToListAsync();
        return Ok(new { logs });
    }
}
